pkg/bgpv1/agent/controller.go

```diff
@@ -0,0 +1,364 @@
+	// The current IPv6 address of the agent, reachable externally.
+	IPv6 net.IP
+}
+
+// Controller is the agent side BGP Control Plane controller.
+//
+// Controller listens for events and drives BGP related sub-systems
+// to maintain a desired state.
+type Controller struct {
+	// NOTE: Swap the usage of Node here to CiliumNode
+	// once #18609 merges
+	//
+	// NodeInformer is a shared informer for Node
+	// resources
+	NodeInformer cache.SharedIndexInformer
+	// NodeLister provides cached and indexed lookups for
+	// Node API objects.
+	NodeLister v1.NodeLister
+	// BGPPolicyInformer is a shared informer for CiliumBGPPeerinPolicy
+	BGPPolicyInformer cache.SharedIndexInformer
+	// PolicyLister provides cached and indexed lookups of
+	// for CilumBGPPeeringPolicy API objects.
+	PolicyLister v2alpha1.CiliumBGPPeeringPolicyLister
+	// Sig informs the Controller that a Kubernetes
+	// event of interest has occurred.
+	//
+	// The signal itself provides no other information,
+	// when it occurs the Controller will query each
+	// informer for the latest API information required
+	// to drive it's control loop.
+	Sig Signaler
+	// BGPMgr is an implementation of the BGPRouterManager interface
+	// and provides a declarative API for configuring BGP peers.
+	BGPMgr BGPRouterManager
+}
+
+// ControllerOpt is a signature for defining configurable options for a
+// Controller
+type ControllerOpt func(*Controller)
+
+// NewController constructs a new BGP Control Plane Controller.
+//
+// When the constructor returns the Controller will be actively watching for
+// events and configuring BGP related sub-systems.
+//
+// The constructor requires an implementation of BGPRouterManager to be provided.
+// This implementation defines which BGP backend will be used (GoBGP, FRR, Bird, etc...)
+// NOTE: only GoBGP currently implemented.
+//
+// Cancel the provided CTX to stop the Controller.
+func NewController(ctx context.Context, rtMgr BGPRouterManager, opts ...ControllerOpt) (*Controller, error) {
+	// create the signaler we will use
+	sig := NewSignaler()
+
+	// informers setup
+	clientset := k8s.CiliumClient()
+	factory := externalversions.NewSharedInformerFactory(clientset, defaultInformerResync)
+	k8sfactory := informers.NewSharedInformerFactory(k8s.Client(), defaultInformerResync)
+
+	nodeLister := k8sfactory.Core().V1().Nodes().Lister()
+	nodeInformer := k8sfactory.Core().V1().Nodes().Informer()
+	nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sig.Event,
+		UpdateFunc: func(_ interface{}, _ interface{}) { sig.Event(struct{}{}) },
+		DeleteFunc: sig.Event,
+	})
+
+	policyLister := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Lister()
+	policyInformer := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Informer()
+	policyInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sig.Event,
+		UpdateFunc: func(_ interface{}, _ interface{}) { sig.Event(struct{}{}) },
+		DeleteFunc: sig.Event,
+	})
+
```

Comment:
@christarazi this is following a best practice for writing k8s controllers. 

Be aware, these are *shared* watchers. Therefore we dont duplicate resources. Shared watchers are specifically designed for decoupled pieces of code to obtain watchers but multiplex over the same connection to the API server (not spam more) 

I believe this pattern removes any "head of line blocking" problems which can occur in the our own watchers infra. Also, it conforms to most other operator/controller code and its use is suggested in various docs explaining the subject.

Reply:
:ok_hand:, that makes sense to me going forward. But I don't see how these shared informers are sharing the cache with the informers created in `pkg/k8s/watchers`, as those are being called via `NewInformer()` which I imagine allocates a whole new object, rather than share it.

Also, out of curiosity, if we convert all the code to shared informers, who / what is creating "the one true" source or cache?

Reply:
I'm not 100% sure if shared informer construction takes account for `NewInformer()` construction. This would be interesting to know. 

> Also, out of curiosity, if we convert all the code to shared informers, who / what is creating "the one true" source or cache?

In the case where all the code uses shared informers, I believe the client-go library will handle the underlying sharing of the cache. I believe the design there is to allow client code not to think much about this, and just instantiate the listers/watchers they require, but ensure lookups are indexed and connections do not sprawl.

Reply:
> In the case where all the code uses shared informers, I believe the client-go library will handle the underlying sharing of the cache. I believe the design there is to allow client code not to think much about this, and just instantiate the listers/watchers they require, but ensure lookups are indexed and connections do not sprawl.

I can start a PR to see how easy this is.

Reply:
I think for the foreseeable future, we will have two implementations in the code at once. One where we use https://github.com/cilium/cilium/issues/15471 and one where we use what's implemented in this PR, because we can't just do a mass refactor all at once, without a major risk of regression. I would like one way to do things because it will make it easier when developing a new feature, to simply pattern match what the current code does.

For now, I want to make sure we aren't duplicating memory hence looking into moving to shared informers everywhere, because we've had numerators reports of OOMs with Cilium and I see this as exacerbating it. Although in this case, BGP is behind a feature flag so the risk is mitigated.